### PR TITLE
Register YAML on class creation

### PR DESCRIPTION
**Pull request by nstarman**
Auto-registration of Cosmology classes with YAML.
Optionally turn off the registration for e.g. temporary classes.

### CODE REVIEW

File: astropy/cosmology/core.py [hunk 1]
```diff
@@ -94,8 +112,8 @@ class Cosmology(metaclass=abc.ABCMeta):
 
     # ---------------------------------------------------------------
 
-    def __init_subclass__(cls):
-        super().__init_subclass__()
+    def __init_subclass__(cls, *, skip_registration: bool = False):
+        _init_subclass(super().__init_subclass__, skip_registration)
```
Comment (mhvk) [inline astropy/cosmology/core.py:116]: This function call seems very strange. Here, you do know that `skip_registration` is not used by the super class (and if anybody creates a new class in between, that's their problem), so this doesn't seem to be needed, i.e., why not have
```
super().__init_subclass__()
```

File: astropy/cosmology/core.py [hunk 2]
```diff
@@ -125,9 +143,18 @@ def __init_subclass__(cls):
...
         # -------------------
-        # register as a Cosmology subclass
+        # Register as a Cosmology subclass
+
+        if skip_registration:
+            return
+
         _COSMOLOGY_CLASSES[cls.__qualname__] = cls
 
+        # register YAML
```
Comment (mhvk) [inline astropy/cosmology/core.py:153]: Not a real issue, but I think I would move all the registration to a new class method, say `_register_class`. Indeed, with that a simpler solution to the problem here is just to override
```
@classmethod
def _register_class(cls):
    pass
```

**Review by mhvk — COMMENTED** [review]
fly-by comments

File: astropy/cosmology/core.py (hunk 2 above)
Comment (nstarman) [reply astropy/cosmology/core.py]: I like this suggestion! Implementing now.

**Comment by nstarman** [comment]
Changelog?

File: astropy/cosmology/core.py [hunk 3]
```diff
@@ -76,6 +76,10 @@ class Cosmology(metaclass=abc.ABCMeta):
 
     For details on how to create performant custom subclasses, see the
     documentation on :ref:`astropy-cosmology-fast-integrals`.
+
+    This class accepts a class keyword-argument ``skip_registration`` (default
```
Comment (mhvk) [inline astropy/cosmology/core.py:80]: Don't forget to remove this addition to the docstring!  Or, maybe better, replace it with something that states that by default classes are registered, and one can avoid that by defining a no-op `_register_cls` (or do that in the `_register_cls` docstring?)

**Review by mhvk — COMMENTED** [review]
Seeing it in practice, yes, this looks good and simple!

File: astropy/cosmology/core.py (hunk 3 above)
Comment (nstarman) [reply astropy/cosmology/core.py]: Done!

**Review by mhvk — APPROVED** [review]
Looks good!

**Comment by pllim** [comment]
Are you sure this does not need a change log? Sounds like a new feature.

**Comment by nstarman** [comment]
The only way it's different is that user-defined classes are auto-registered, as opposed to only the pre-built classes. So maybe a changelog is warranted?

**Comment by pllim** [comment]
> The way it's different is that user-defined classes are auto-registered, as opposed to only the pre-built classes. So maybe a changelog is warranted?

Yes, I think so. Can you please open a follow up PR with the change log? Thanks!

**Comment by pllim** [comment]
I also think this broken pyinstaller job https://github.com/astropy/astropy/actions/runs/5373414515/jobs/9747769484

**Comment by pllim** [comment]
If I don't see a solution in a reasonable time frame, I am going to propose reverting this PR. FYI. Also see https://github.com/astropy/astropy/issues/14986 . Thanks!

**Pull request merged**